test_client/src/main/java/io/reactivesocket/tck/ParseMarble.java

```diff
@@ -35,117 +39,101 @@ public ParseMarble(String marble, Subscriber<? super Payload> s) {
-    public void parse(long n) {
-
-        long numSent = 0;
-        // if cancel has been called, don't do anything
-        if (cancelled) return;
-
-        String buffer = "";
-        boolean grouped = false;
+    // this is for channel, when the marble can be added incrementally
+    public ParseMarble(Subscriber<? super Payload> s) {
+        this.s = s;
+        this.marble = "";
+        parseLatch = new CountDownLatch(1);
+        sendLatch = new CountDownLatch(1);
+    }
 
-        while (iter.current() != StringCharacterIterator.DONE) {
-            if (numSent >= n) return;
-            char c = iter.current();
-            switch (c) {
-                case '-':
-                    if (grouped) buffer += c;
-                    else try {Thread.sleep(10);} catch (Exception e) {System.out.println("Interrupted");}
-                    break;
-                case '|':
-                    if (grouped) buffer += c;
-                    else s.onComplete();
-                    break;
-                case '#':
-                    if (grouped) buffer += c;
-                    else s.onError(new Throwable("error"));
-                    break;
-                case '(':
-                    // ignore groupings for now
-                    break;
-                case ')':
-                    // ignore groupings for now
-                    break;
-                default:
-                    if (argMap != null) {
-                        // this is hacky, but we only expect one key and one value
-                        Map<String, String> tempMap = argMap.get(c + "");
-                        if (tempMap == null) {
-                            s.onNext(new PayloadImpl(c + "", c + ""));
-                            break;
-                        }
-                        List<String> key = new ArrayList<>(tempMap.keySet());
-                        List<String> value = new ArrayList<>(tempMap.values());
-                        s.onNext(new PayloadImpl(key.get(0), value.get(0)));
-                    }
-                    else s.onNext(new PayloadImpl(c + "", c + "")); // if value not mapped, just send as data and meta
-                    numSent++;
-                    break;
-            }
-            iter.next();
-        }
+    // adds stuff to the end of marble
+    public synchronized void add(String m) {
+        System.out.println("adding " + m);
+        this.marble += m;
+        parseLatch.countDown();
     }
 
-    /*private void parseMarble(String marble, Subscriber<? super Payload> s) {
-        String buffer = "";
-        boolean grouped = false;
+    public synchronized void request(long n) {
+        System.out.println("requested" + n);
+        numRequested += n;
+        if (marble.length() > marbleIndex) sendLatch.countDown();
+    }
 
+    // this parses the actual marble diagram and acts out the behavior
+    // should be called upon triggering a handler
+    public void parse() {
```

Comment:
I wouldn't say step(), since it doesn't move just one element forward, it's more like a `run()`, as this method is being called by a thread who's only job is to run the loop until it returns.